File: backend/app/api/react.py
# Built with Spec4 AI - https://spec4.ai
"""The ReAct loop app's HTTP surface: the presets, the run stream, the trace.

`GET /presets` serves the curated catalogue from module constants -- no model,
no database, no quota, nothing the visitor supplies. Served from the API rather
than duplicated in the frontend for the same reason as the orchestrated roster
and the collaboration identity cards: the question a run actually asks is
resolved server-side from these constants, so a frontend copy would be a second
source of truth free to drift from the one the loop uses. **It carries no
answer to any preset, because the catalogue holds none** -- see
`react/presets.py`.

`POST /run` streams the run, one envelope per event. It is a stream rather than
a JSON response because the whole point of the demonstration is watching the
loop decide: a thought, then an action, then the observation it produced, then
the next thought built on it. A response assembled at the end would show a
visitor the result of an interleaved loop with the interleaving removed.

`GET /run/{run_id}` reads one completed trace back whole. It is what the
client's stored run ids are for: there is no server-side visitor identity here,
so a visitor returning to the page re-fetches their trace by id rather than
being trusted to have cached it correctly.

**This router is thin on purpose.** The catalogue projection, the stream and
the persistence live in the slice service; what is here is the request body,
the event encoding, the session lifetime and the disconnect handling.

Like the planning, orchestrated and collaboration run endpoints, `/run`
deliberately does **not** use `Depends(get_db_session)`: a dependency's session
is bound to the request scope and this response outlives the handler, because
the generator is still running -- and still writing the run's row -- while the
body streams. The session is opened inside the generator instead.
"""

from __future__ import annotations

import asyncio
import json
import uuid
from collections.abc import AsyncGenerator

import structlog
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

# Imported from the package root rather than `sse_starlette.sse`: only the root
# declares these in `__all__`, and this module is type-checked strictly (unlike
# the pre-v6 routers, which inherit an exemption).
from sse_starlette import EventSourceResponse, ServerSentEvent

from backend.app.db.session import async_session_factory, get_db_session
from backend.app.react import service, suitability
from backend.app.react.presets import PRESET_QUESTIONS
from backend.app.react.schemas import (
    PresetsResponse,
    QuestionSuitability,
    RunRequest,
    TraceResponse,
)
from backend.app.services import text_gate
from backend.app.services.moderation import (
    Moderator,
    get_moderator,
    get_stateless_moderator,
)

logger = structlog.get_logger()

router = APIRouter(prefix="/api/react", tags=["react"])

#: Seconds between keep-alive comments, matching `api/collab.py`. Render's
#: proxy closes an idle connection, and a cycle that is waiting on a model and
#: then on a search can run for tens of seconds with nothing to say.
PING_SECONDS = 15


@router.get("/presets")
async def get_presets() -> PresetsResponse:
    """Return the five curated multi-hop questions.

    Static and free: no model call, no quota, and nothing a visitor supplied.
    It answers on a deployment with no provider keys at all, which is what
    makes it safe to fetch on page load.

    Returns:
        The catalogue's questions and display metadata, and the run's
        server-fixed search-cycle budget. Never an answer to any of them.
    """
    return service.public_presets()


def _event(name: str, payload: dict[str, object]) -> ServerSentEvent:
    """Encode one envelope as a named SSE event.

    Args:
        name: The event name the client listens for.
        payload: JSON-serialisable body.

    Returns:
        The encoded event.
    """
    return ServerSentEvent(event=name, data=json.dumps(payload))


@router.post("/run")
async def run(
    payload: RunRequest,
    moderator: Moderator = Depends(get_stateless_moderator),
) -> EventSourceResponse:
    """Start a run and stream one envelope per event.

    A 200 with any refusal carried as an `error` event, following the
    convention the planning, orchestrated and collaboration apps set: a run
    that produced cycles and then stopped must not push the client's error
    branch and discard them.

    A free-form question passes the shared moderation gate first; a curated
    preset does not, because the gate recognises the app's own canonical text
    rather than accepting an id. `get_stateless_moderator` rather than the
    session-bound one, for the same reason planning's `/run` uses it: this
    response outlives its handler, so it must not hold a request-scoped
    session. The cost is one `moderation_log` row, which is the documented
    trade for a streaming route.

    Args:
        payload: The validated request body. Exactly one of
            `preset_question_id` and `visitor_question` is present.
        moderator: The shared safety gate.

    Returns:
        The event stream.
    """
    return EventSourceResponse(_run_stream(payload, moderator), ping=PING_SECONDS)


async def _run_stream(
    payload: RunRequest, moderator: Moderator
) -> AsyncGenerator[ServerSentEvent, None]:
    """Drive the run and put each envelope on the wire as it is produced.

    Args:
        payload: The validated request body.
        moderator: The shared safety gate, injected so a test can substitute it.

    Yields:
        One SSE event per envelope.
    """
    run_id = uuid.uuid4()
    delivered = 0

    # The shared gate, before anything is reserved or spent. A curated preset
    # never reaches it: `PRESET_QUESTIONS` is the server's own canonical text
    # and the gate byte-matches against it, so a preset id is never a claim that
    # buys a bypass. The gate costs no model allowance of its own, which is what
    # lets it run first.
    if payload.visitor_question is not None:
        # A session of its own, opened and closed before the stream begins. The
        # streaming rule forbids a *request-scoped* session here, not any
        # session -- and without one the gate's verdict would go unrecorded for
        # a client that starts a run without checking the question first.
        async with async_session_factory() as gate_session:
            gate = await text_gate.check_free_text(
                payload.visitor_question,
                app_name=service.REACT_APP_NAME,
                curated=PRESET_QUESTIONS,
                session=gate_session,
                moderator=moderator,
            )
        if not gate.allowed and gate.code is not None:
            yield _event(
                "error",
                {
                    "code": gate.code,
                    "message": gate.message or "",
                    "stub": False,
                },
            )
            logger.info("react_run_refused", run_id=str(run_id), reason=gate.code)
            return

    # Reused from the cache, never re-asked. The check already ran while the
    # visitor was typing; spending a second call to record the same verdict on
    # the run row would be paying twice for one answer. A cache miss records
    # nothing, which is correct -- the columns mean "this is what the check
    # said", not "a check was attempted".
    verdict = (
        suitability.cached(payload.visitor_question)
        if payload.visitor_question is not None
        else None
    )

    async with async_session_factory() as session:
        stream = service.stream_run(
            session, run_id=run_id, request=payload, suitability=verdict
        )
        try:
            async for event in stream:
                delivered += 1
                yield _event(event.name, event.payload)
        except asyncio.CancelledError:
            # Abandoning an `async for` does not close the inner generator, so
            # the loop would keep cycling against a stream nobody reads --
            # spending model and search quota on a run the visitor walked away
            # from, which matters most here because this is the gallery's most
            # expensive example per run. Closing it explicitly is what makes
            # the stop immediate, and it is what runs the generator's `finally`
            # so the run's unspent reservation is given back.
            await stream.aclose()
            logger.info("react_run_abandoned", run_id=str(run_id), events=delivered)
            raise

    logger.info("react_run_stream_closed", run_id=str(run_id), events=delivered)


class SuitabilityRequest(BaseModel):
    """Body of `POST /api/react/suitability`.

    Attributes:
        visitor_question: The visitor's own question. Presets never reach here.
        session_id: The browser session, for the per-session check cap.
    """

    visitor_question: str = Field(
        min_length=1, max_length=suitability.MAX_QUESTION_CHARS
    )
    session_id: str = Field(min_length=1, max_length=64)


class SuitabilityResponse(BaseModel):
    """Response body of `POST /api/react/suitability`.

    `verdict` is null for the neutral state, which is what every failure path
    resolves to. **The response is a 200 either way**: an advisory that could
    not be produced is not an error the visitor needs to see, and returning a
    5xx would push the client's error branch for a hint.

    Attributes:
        verdict: The assessment, or null when nothing could assess it.
        checks_remaining: How many checks this session may still spend, so the
            client can stop asking rather than discovering the cap.
    """

    verdict: QuestionSuitability | None
    checks_remaining: int


@router.post("/suitability")
async def check_suitability(
    payload: SuitabilityRequest,
    moderator: Moderator = Depends(get_moderator),
) -> SuitabilityResponse:
    """Advise whether a free-form question will exercise the loop.

    **This never blocks a run, and it never spends the visitor's two-run
    allowance.** It costs one call against the shared framework cap and nothing
    else; a refusal, a timeout or an exhausted chain all resolve to a null
    verdict the UI renders as a soft note.

    The moderation gate still applies: a question refused here is refused, and
    the refusal is reported as the same status the rest of the gallery uses so
    a visitor meets one treatment rather than six. Unlike `/run`, this is an
    ordinary request-scoped route, so it takes the **session-bound** moderator
    and the verdict lands in `moderation_log` -- a salted hash, the category and
    the latency, and never the question.

    Args:
        payload: The question and the session id.
        moderator: The shared safety gate.

    Returns:
        The verdict or the neutral state, plus the session's remaining checks.

    Raises:
        HTTPException: 422 when the question was refused, 503 when nothing
            could examine it. The two must stay distinguishable: one the
            visitor fixes by rewording, the other they cannot fix at all.
    """
    gate = await text_gate.check_free_text(
        payload.visitor_question,
        app_name=service.REACT_APP_NAME,
        curated=PRESET_QUESTIONS,
        moderator=moderator,
    )
    if not gate.allowed and gate.code is not None:
        raise HTTPException(
            status_code=text_gate.status_for(gate.code),
            detail=gate.message or "",
        )

    verdict = await suitability.assess(
        payload.visitor_question, session_id=payload.session_id
    )
    return SuitabilityResponse(
        verdict=verdict,
        checks_remaining=suitability.checks_remaining(payload.session_id),
    )


@router.get("/run/{run_id}")
async def get_run(
    run_id: str,
    session: AsyncSession = Depends(get_db_session),
) -> TraceResponse:
    """Return one completed run's trace, whole.

    A request-scoped session is correct here, unlike on `/run`: this response
    is assembled and returned before the handler exits.

    Args:
        run_id: The run's id, as a UUID string.
        session: The request-scoped database session.

    Returns:
        The stored trace.

    Raises:
        HTTPException: 404 when no run carries that id, including when the id
            is not a well-formed UUID -- an unparseable id names no run, which
            is the same answer.
    """
    try:
        parsed = uuid.UUID(run_id)
    except ValueError:
        raise HTTPException(status_code=404, detail="No such run.") from None

    trace = await service.load_run(session, parsed)
    if trace is None:
        raise HTTPException(status_code=404, detail="No such run.")
    return trace

File: frontend/src/api/react.ts
// Built with Spec4 AI - https://spec4.ai
import { fetchEventSource } from '@microsoft/fetch-event-source'

/**
 * Typed client for the ReAct loop endpoints.
 *
 * One client per example app, matching `collab.ts`, `orchestrated.ts` and
 * `planning.ts`, including their error messages: `fetch` rejects identically
 * for "the backend isn't running" and "the browser blocked the response for
 * CORS", so the unreachable message names both rather than saying something
 * the reader cannot act on.
 *
 * The presets are fetched rather than duplicated here. The question a run
 * actually asks is resolved server-side from the same constants, so a frontend
 * copy would be a second source of truth free to drift from the one the loop
 * uses — and it would put one question on screen while another went to the
 * model.
 *
 * **There is no `cycleBudget` option on the run request, and none should be
 * added.** The budget is server-fixed at eight search cycles; the design mock's
 * 3–6 select is superseded by the stack spec's `react_run_call_budget`
 * decision. The run reserves its whole worst case before the first cycle, so a
 * client-supplied budget would let a caller reserve one number and spend
 * another.
 */

/** One curated multi-hop question, as the selector sees it. */
export interface ReactPreset {
  id: string
  /** Short chip text. Deliberately does not resolve any of the question's hops. */
  label: string
  /** The question put to the model, verbatim. */
  question: string
  /** How many facts the question chains. */
  hopCount: number
  /** True for the presets curated so every hop needs a live observation. */
  guaranteedFullyObserved: boolean
}

/** Response body of GET /api/react/presets. */
export interface ReactPresetsResponse {
  presets: ReactPreset[]
  setVersion: string
  /** The server-fixed search-cycle ceiling. Published so nothing hardcodes it. */
  cycleBudget: number
}

/** One search result, rendered verbatim as part of a cycle's observation. */
export interface ObservationSnippet {
  /** 1-based position within its observation, as the provider ranked them. */
  idx: number
  title: string
  snippet: string
  url: string
  /** Rendered as "undated" when null — an empty slot reads as "recent". */
  published_date: string | null
  /** True when the snippet was cut to fit the model's context. */
  truncated: boolean
}

/**
 * What happened when a query was issued.
 *
 * Three values, not two, and the third is why the UI needs this: an empty
 * result and an unreachable provider both show no snippets and are different
 * facts — one is the web's answer to the question, the other is the
 * demonstration failing. They get different visible states.
 */
export type ObservationStatus = 'ok' | 'empty' | 'unavailable'

/** Whether the final answer's citations point at observations that exist. */
export interface GroundingAudit {
  all_cited_present: boolean
  cited: number[]
  unverified: number[]
}

/** Where one hop's fact came from, after the server's cross-checks. */
export interface HopAnnotation {
  cycle_index: number
  fact: string
  source: 'observation' | 'model_knowledge' | 'mixed'
  /** The cycle whose observation supplies it. Null once downgraded. */
  supporting_cycle: number | null
  note: string
}

/**
 * The annotation panel's payload.
 *
 * `all_hops_observed` is **derived on the server** from the annotations that
 * survived its cross-checks — the model has no field to assert it with. The
 * panel may therefore state it; it could not if the model had claimed it.
 */
export interface AnnotationResult {
  hops: HopAnnotation[]
  all_hops_observed: boolean
  observed_count: number
  recalled_count: number
  dropped: string[]
  downgraded: number[]
}

/** Why a run ended without an answer. */
export type ExhaustionReason =
  | 'search_ceiling'
  | 'malformed_step'
  | 'search_unavailable'
  | 'model_unavailable'
  | 'wall_clock'
  | 'call_budget' 

/**
 * An event from `POST /api/react/run`.
 *
 * Discriminated on `kind`, which is also the SSE event name. Anything the
 * client does not recognise is ignored rather than crashed on, so the backend
 * can add an event type without breaking a deployed frontend.
 *
 * Thought, action, observation and counter are four events for one cycle,
 * deliberately: they are separated in time by seconds, and the order is the
 * thing the app exists to show.
 */
export type ReactRunEvent =
  | {
      kind: 'run_started'
      run_id: string
      question: string
      question_source: 'preset' | 'custom'
      preset_id: string | null
      cycle_budget: number
      runs_remaining: number
      stub: boolean
    }
  | { kind: 'cycle_thought'; cycle: number; thought: string; stub: boolean }
  | {
      kind: 'cycle_action'
      cycle: number
      action_kind: 'search' | 'answer'
      query: string | null
      rationale: string
      stub: boolean
    }
  | {
      kind: 'cycle_observation'
      /** 1-based observation number — the value a final answer cites. */
      index: number
      /** The exact query issued, verbatim. */
      query: string
      results: ObservationSnippet[]
      is_empty: boolean
      status: ObservationStatus
      /** Why the search could not run. Null unless `status` is `unavailable`. */
      detail: string | null
      truncated: boolean
      stub: boolean
    }
  | { kind: 'cycle_counter'; searches_used: number; cycle_budget: number; stub: boolean }
  | {
      kind: 'final_answer'
      run_id: string
      answer: string
      observation_cycles: number[]
      audit: GroundingAudit
      searches_used: number
      cycle_budget: number
      stub: boolean
    }
  | {
      kind: 'budget_exhausted'
      run_id: string
      reason: ExhaustionReason
      unresolved: string[]
      partial_findings: number[]
      searches_used: number
      cycle_budget: number
      stub: boolean
    }
  | ({ kind: 'hop_annotations' } & AnnotationResult)
  | { kind: 'error'; code: string; message: string; stub: boolean }

/** The event names the server sends. Anything else is ignored, not crashed on. */
const RUN_EVENT_NAMES = new Set([
  'run_started',
  'cycle_thought',
  'cycle_action',
  'cycle_observation',
  'cycle_counter',
  'final_answer',
  'budget_exhausted',
  'hop_annotations',
  'error',
])

/** The three ways a stream stops. Exactly one arrives, and it is last. */
export const TERMINAL_EVENT_KINDS = new Set(['final_answer', 'budget_exhausted', 'error'])

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL ?? 'http://localhost:8000'

/** A failed ReAct request, carrying the backend's own explanation where there was one. */
export class ReactRequestError extends Error {
  readonly code: string

  constructor(message: string, code: string, cause?: unknown) {
    super(message, { cause })
    this.name = 'ReactRequestError'
    this.code = code
  }
}

/**
 * `fetch` rejects identically whether the backend is down or the browser
 * blocked the response for CORS, so the message names both.
 */
function unreachable(url: string, cause: unknown): ReactRequestError {
  return new ReactRequestError(
    `Could not reach the backend at ${url}. Either it isn't running, or this page's origin (${window.location.origin}) doesn't match the backend's CORS_ORIGIN setting.`,
    'unreachable',
    cause,
  )
}

/** The wire shape of one preset, before it is camel-cased for the components. */
interface PresetWire {
  id: string
  label: string
  question: string
  hop_count: number
  guaranteed_fully_observed: boolean
}

interface PresetsWire {
  presets: PresetWire[]
  set_version: string
  cycle_budget: number
}

/**
 * Fetch the five curated multi-hop questions.
 *
 * Static server-side configuration: no model call, no quota, nothing the
 * visitor supplied, and — because the catalogue stores questions only — no
 * answer to any of them. It answers on a deployment with no provider keys at
 * all, which is what makes it safe to fetch on page load.
 *
 * @returns The presets, the catalogue version, and the run's cycle budget.
 * @throws ReactRequestError if the backend is unreachable or returns non-2xx.
 */
export async function fetchReactPresets(): Promise<ReactPresetsResponse> {
  const url = `${API_BASE_URL}/api/react/presets`

  let response: Response
  try {
    response = await fetch(url)
  } catch (cause) {
    throw unreachable(url, cause)
  }

  if (!response.ok) {
    throw new ReactRequestError(
      `The backend returned ${response.status} for ${url}.`,
      'request_failed',
    )
  }

  const body = (await response.json()) as PresetsWire
  return {
    presets: body.presets.map((preset) => ({
      id: preset.id,
      label: preset.label,
      question: preset.question,
      hopCount: preset.hop_count,
      guaranteedFullyObserved: preset.guaranteed_fully_observed,
    })),
    setVersion: body.set_version,
    cycleBudget: body.cycle_budget,
  }
}

/** The advisory verdict on a visitor's own question. */
export interface QuestionSuitability {
  verdict: 'multi_hop_live' | 'multi_hop_static' | 'single_hop' | 'unanswerable'
  estimated_hops: number
  requires_live_info: boolean
  live_hop_description: string | null
  exercises_loop: boolean
  confidence: 'low' | 'medium' | 'high'
  /** One plain sentence, already sanitised server-side. Safe to render verbatim. */
  visitor_message: string
}

/** Response body of POST /api/react/suitability. */
export interface SuitabilityResponse {
  /** Null is the neutral "unknown" state — every failure path resolves to it. */
  verdict: QuestionSuitability | null
  checks_remaining: number
}

/** Longest question the server will accept. Enforced client-side too. */
export const MAX_QUESTION_CHARS = 300

/**
 * Ask whether a free-form question will exercise the loop.
 *
 * **Advisory only.** The caller must never let the result — or its absence, or
 * a failure of this call — decide whether Start is enabled. A null `verdict` is
 * the ordinary neutral state, not an error: the server resolves a timeout, an
 * exhausted model chain, a spent session cap and a twice-invalid response all
 * to the same thing, because the visitor's response to each is identical.
 *
 * A moderation refusal is the one case that *is* an error, and it arrives as a
 * 422 (reword it) or a 503 (nothing could check it) — kept apart because one
 * the visitor can fix and the other they cannot.
 *
 * @param question - The visitor's own question.
 * @param sessionId - The browser session, for the server's per-session check cap.
 * @returns The verdict or the neutral state, plus remaining checks.
 * @throws ReactRequestError on a moderation refusal or an unreachable backend.
 */
export async function checkSuitability(
  question: string,
  sessionId: string,
): Promise<SuitabilityResponse> {
  const url = `${API_BASE_URL}/api/react/suitability`

  let response: Response
  try {
    response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ visitor_question: question, session_id: sessionId }),
    })
  } catch (cause) {
    throw unreachable(url, cause)
  }

  if (!response.ok) {
    const body: unknown = await response.json().catch(() => null)
    const detail =
      typeof body === 'object' && body !== null && 'detail' in body
        ? String((body as { detail: unknown }).detail)
        : `The backend returned ${response.status} for ${url}.`
    throw new ReactRequestError(
      detail,
      response.status === 422 ? 'moderation_blocked' : 'moderation_unavailable',
    )
  }

  return (await response.json()) as SuitabilityResponse
}

/** One stored cycle, as `GET /api/react/run/{run_id}` returns it. */
export interface StoredCycle {
  cycle: number
  thought: string
  action: { kind: 'search' | 'answer'; query: string | null }
  observation: {
    index: number
    query: string
    results: ObservationSnippet[]
    is_empty: boolean
    status: ObservationStatus
    detail: string | null
  } | null
}

/** A completed run's whole persisted trace. */
export interface StoredTrace {
  run_id: string
  created_at: string
  question_origin: string
  searches_used: number
  cycle_budget: number
  ending: 'final_answer' | 'budget_exhausted' | null
  duplicate_queries_blocked: number
  empty_observations: number
  cycle_trace: StoredCycle[]
  terminal_card: Record<string, unknown> | null
}

/**
 * Fetch one completed run's trace by id.
 *
 * **Read from the server rather than from the browser's cached copy**, which is
 * why only the run id is stored locally. The persisted record is the one the
 * grounding audit was computed against; a cached trace could drift from it and
 * would then be showing a visitor something no run actually produced.
 *
 * @param runId - The run's id, from the local allowance record.
 * @returns The whole stored trace.
 * @throws ReactRequestError if the backend is unreachable or has no such run.
 */
export async function fetchReactRun(runId: string): Promise<StoredTrace> {
  const url = `${API_BASE_URL}/api/react/run/${runId}`

  let response: Response
  try {
    response = await fetch(url)
  } catch (cause) {
    throw unreachable(url, cause)
  }

  if (!response.ok) {
    throw new ReactRequestError(
      `The backend returned ${response.status} for ${url}.`,
      response.status === 404 ? 'not_found' : 'request_failed',
    )
  }

  return (await response.json()) as StoredTrace
}

/** Options for {@link startReactRun}. Exactly one question source, as the server requires. */
export interface StartReactRunOptions {
  presetQuestionId?: string
  visitorQuestion?: string
  sessionId: string
  onEvent: (event: ReactRunEvent) => void
  signal?: AbortSignal
}

/**
 * Start a run and deliver each envelope as it arrives.
 *
 * `@microsoft/fetch-event-source` rather than the browser's `EventSource`, and
 * that is not a preference: `EventSource` is GET-only and a run starts from a
 * POST body. **Do not add a GET variant of the endpoint** to work around it —
 * that forks the API for a transport detail.
 *
 * Three of the library's defaults are wrong for an endpoint that spends quota,
 * and all three are overridden here for the reasons the planning, orchestrated
 * and collaboration clients override them. They matter more here than anywhere
 * else in the gallery, because this is its most expensive example per run:
 *
 * - `openWhenHidden: true`, because the default drops the connection on a
 *   backgrounded tab and *reopens* it — paying for the run twice, silently, and
 *   only for visitors who switch tabs.
 * - `onopen` throws the backend's own explanation, so a 422 on the request body
 *   does not read as a transport fault.
 * - `onerror` rethrows. The default retries forever whenever it *returns*.
 *
 * @param options - The question source, the session id, and the event sink.
 * @returns A promise resolving when the stream closes.
 * @throws ReactRequestError if the backend is unreachable or rejects the request.
 */
export async function startReactRun({
  presetQuestionId,
  visitorQuestion,
  sessionId,
  onEvent,
  signal,
}: StartReactRunOptions): Promise<void> {
  const url = `${API_BASE_URL}/api/react/run`

  await fetchEventSource(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Accept: 'text/event-stream' },
    body: JSON.stringify({
      preset_question_id: presetQuestionId ?? null,
      visitor_question: visitorQuestion ?? null,
      session_id: sessionId,
    }),
    signal,
    openWhenHidden: true,

    async onopen(response) {
      if (response.ok) {
        return
      }
      const body: unknown = await response.json().catch(() => null)
      const detail =
        typeof body === 'object' && body !== null && 'detail' in body
          ? String((body as { detail: unknown }).detail)
          : `The backend returned ${response.status} for ${url}.`
      throw new ReactRequestError(detail, 'request_failed')
    },

    onmessage(message) {
      if (!RUN_EVENT_NAMES.has(message.event)) {
        return
      }
      const payload = JSON.parse(message.data) as Record<string, unknown>
      // `kind` is the SSE event name, and the action's own `kind` field would
      // collide with it — so the action's is renamed on the way in rather than
      // one of them silently winning.
      if (message.event === 'cycle_action') {
        const { kind: actionKind, ...rest } = payload
        onEvent({ kind: 'cycle_action', action_kind: actionKind, ...rest } as ReactRunEvent)
        return
      }
      onEvent({ kind: message.event, ...payload } as ReactRunEvent)
    },

    onerror(cause) {
      if (cause instanceof ReactRequestError) {
        throw cause
      }
      throw unreachable(url, cause)
    },
  })
}
